.github/workflows/trivy.yml

```diff
@@ -0,0 +1,87 @@
+---
+#
+# This workflow runs Trivy on a Docker image
+# It can pull the image from a container registry
+# or download a tar file.  The latter is used
+# to check a container image prior to publishing
+# to the registry.
+
+name: Run Trivy on a Docker image and push results to GitHub
+
+on:
+  workflow_call:
+    inputs:
+      SOURCE_TYPE: # 'tar' or 'image'
+        required: true
+        type: string
+      TARFILE_NAME: # only used if SOURCE_TYPE=='tar'
+        required: false
+        type: string
+      IMAGE_NAME:
+        required: true
+        type: string
+      EXIT_CODE: # return code for failed scan. 0 means OK
+        required: false
+        type: number
+        default: 0
+
+env:
+  sarif_file_name: trivy-results.sarif
+
+jobs:
+  trivy:
+    name: Trivy
+    runs-on: ubuntu-latest
+    permissions:
+      contents: read
+      security-events: write
+
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v4
+
+      - name: Download tar file
+        id: tar-download
+        uses: actions/download-artifact@v4
+        if: ${{ inputs.SOURCE_TYPE == 'tar' }}
+        with:
+          name: ${{ inputs.TARFILE_NAME }}
+          path: /tmp
+
+      - name: load docker image from tar file
+        if: ${{ inputs.SOURCE_TYPE == 'tar' }}
+        run: cat ${{ steps.tar-download.outputs.download-path
+          }}/${{ inputs.TARFILE_NAME
+          }} | docker import - ${{ inputs.IMAGE_NAME }}
+
+      - name: Run Trivy vulnerability scanner for any major issues
+        uses: aquasecurity/trivy-action@0.24.0
+        id: trivy
+        with:
+          image-ref: ${{ inputs.IMAGE_NAME }}
+          ignore-unfixed: true # skip vul'ns for which there is no fix
+          # list files to skip, each with a justification
+          skip-files: |
+            /usr/local/lib/R/site-library/gargle/extdata/fake_service_account.json
+            /usr/local/lib/R/site-library/openssl/doc/keys.html
+          # fake_service_account.json is a fake account that gets
+          #  flagged as a credentials file
+          # keys.html is a documentation file that appears to contain cred's
+          severity: 'CRITICAL,HIGH'
+          format: 'sarif'
+          # only output findings for configured severities
+          limit-severities-for-sarif: true
+          output: ${{ env.sarif_file_name  }}
+          exit-code: ${{ inputs.EXIT_CODE }}
+
+      - name: Upload Trivy scan results to GitHub Security tab
+        uses: github/codeql-action/upload-sarif@v3.25.12
+        # This is the recommended way to upload scan results
+        # after Trivy exits with HIGH/CRITICAL findings
+        # See https://github.com/aquasecurity/trivy-action?\
+        # tab=readme-ov-file#using-trivy-with-github-code-scanning
+        if: ${{ success() || steps.trivy.conclusion=='failure' }}
```

Comment:
you have not address this concern?

Reply:
Weird:  I typed a long answer but it seems to have been lost...

In short the conditional will evaluate to `true` if either (1) all preceding steps passed or (2) all steps EXCEPT the trivy step passed. It will fail if any other earlier step (the download-tar or load-docker-from-tar) failed.   The linked page suggests using `always()` which doesn't quite do the right thing.

Reply:
> The linked page suggests using always() which doesn't quite do the right thing.

If this is a workaround for using `always()` then could you change the comments to reflect that?

> If, say, the download-tar step fails, then we want the whole workflow to fail.

This doesn't seem to be true for  condition success()=F or steps.trivy.conclusion=='failure'?  Doesn't that mean upload whether or not any previous step fails?

Reply:
It's not a "workaround" but an improvement but, yes, I will add a comment to explain the logic for not using `always()`.  

> Doesn't that mean upload whether or not any previous step fails?
No:  If, say, the download-tar step fails then success()=False and steps.trivy.conclusion='skipped', so the `if:` expression will evaluate to False.

Reply:
what about the condition where.. download-tar step fails then success()=False and steps.trivy.conclusion='failure'?  that seems like it would NOT result in a workflow failure and still attempt to upload trivy results?

Reply:
@zaro0508 

> what about the condition where.. download-tar step fails then success()=False and steps.trivy.conclusion='failure'?

That can never happen:  If download-tar fails then none of the subsequent steps ("load docker image from tar file" or "Run Trivy vulnerability scanner for any major issues") will run.  And if the trivy step does not run then steps.trivy.conclusion is not 'failure' but rather 'skipped.'  Maybe read up on [steps.*.conclusion](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/contexts#steps-context) ?